### PR TITLE
Implement WSF/GSF satfunc familiy for CO2STORE

### PR DESCRIPTION
**Pull request by totto82**
See test case https://github.com/OPM/opm-tests/pull/902

### CODE REVIEW

**Comment by totto82** [comment]
jenkins build this please

**Comment by totto82** [comment]
jenkins build this please

File: src/opm/input/eclipse/EclipseState/Grid/SatfuncPropertyInitializers.cpp [hunk 1]
```diff
@@ -930,6 +1004,8 @@ namespace {
                 return ph.active(::Opm::Phase::GAS)
                     ? Opm::fun::map( famII_3p, Opm::fun::iota( num_tables ) )
                     : Opm::fun::map( famII_2p, Opm::fun::iota( num_tables ) );
+            case SatfuncFamily::III:
+                throw std::domain_error("Saturation keyword family III is not applicable for a oil systems");
```
Comment (bska) [inline src/opm/input/eclipse/EclipseState/Grid/SatfuncPropertyInitializers.cpp:1008]: Tiny nit: We should not mix singular and plural in the diagnostic message ("A .. systemS").  We either need to say "..for AN oil system" (singular) or "... for oil systems" (plural).

File: src/opm/output/eclipse/Tables.cpp [hunk 2]
```diff
@@ -1164,6 +1244,83 @@ namespace { namespace SatFunc {
...
+                // Krw(Sw)
+                detail::outputRelperm(t.getKrwColumn(), tolcrit,
+                                      linTable.column(tableID, primID, 1));
+
+                // Pcow = zero
+                {
+                    const auto& krw = t.getKrwColumn();
+                    std::transform(std::begin(krw), std::end(krw),
+                                   linTable.column(tableID, primID, 2),
+                                   [&units](const double Pc) -> double
```
[the remark this answers is not in the document]
Comment (bska) [reply src/opm/output/eclipse/Tables.cpp:1311]: > or possibly just remove it altogether.

And by "it" here I mean the parameter name.  The lambda itself needs to have a `double` parameter to satisfy the requirements of `std::transform()`.

**Comment by totto82** [comment]
> Do we guarantee somewhere that we don't have any xSF tables unless both gas and water are active phases? I didn't see any obvious checks for this. I would probably also add some further restriction on the possible saturation function keywords in the context of xSF.

Yes in the diagnostics. https://github.com/OPM/opm-simulators/pull/4485

**Comment by bska** [comment]
> > Do we guarantee somewhere that we don't have any xSF tables unless both gas and water are active phases? I didn't see any obvious checks for this. I would probably also add some further restriction on the possible saturation function keywords in the context of xSF.
> 
> Yes in the diagnostics. OPM/opm-simulators#4485

I see.  In my opinion that check happens too late, though arguably before we actually extract the end points.  On the other hand, those diagnostics [only log](https://github.com/OPM/opm-simulators/blob/05caaa894620220929b1c3179aef36377c93b19d/opm/core/props/satfunc/RelpermDiagnostics.cpp#L160) the inconsistent phases/keywords and does not stop the simulation.  I'd really prefer that we not start the simulation run if we're in that situation.  Switching from `OpmLog::info()` to `OpmLog::error()` is not sufficient to stop the simulation, so we'll need to do more.

**Comment by totto82** [comment]
Thanks for the review. I didn't explicit force the user to only combine it with CO2STORE as I dont see why you cant use WSF and GSF for a standard water and gas case.

**Comment by totto82** [comment]
jenkins build this please

**Review by bska — APPROVED** [review]
Thanks a lot for the updates.  I'll merge this into master now.

**Pull request merged**